### PR TITLE
fix(dynamics): avoid double particle thermostat coupling in NPT

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit Pull Request

## Description

Fix high-level NPT particle velocity half-steps so explicit particle Nose-Hoover scaling is not combined with the NPT velocity primitive's inline thermostat drag. Update the Toolkit NPT barostat bridge to use an eta-free caller contract while zeroing `eta_dots` for legacy ops backends that still require the argument.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes to #95. Relates to #91

## Changes Made

- Route high-level `NPT.pre_update()` and `NPT.post_update()` particle velocity half-steps through `nph_velocity_half_step`.
- Update the `npt_barostat_half_step` bridge documentation and legacy-backend dispatch for eta-free Toolkit callers.
- Add a regression test that asserts high-level NPT uses the no-thermostat velocity primitive.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

Staged whitespace/conflict check passed. Targeted high-level NPT/fused ops repro passed with final velocity diff `2.541e-05 < 1.0e-03`.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

None.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/dynamics/_ops/npt_nph.py [hunk 1]
```diff
@@ -91,6 +93,41 @@
...
+    buf = _ZERO_ETA_DOTS_CACHE.get(key)
+    if buf is None:
+        buf = torch.zeros(
+            cell_velocity.shape[0],
+            1,
+            dtype=cell_velocity.dtype,
+            device=cell_velocity.device,
+        )
+        _ZERO_ETA_DOTS_CACHE[key] = buf
+    return buf
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemi/dynamics/_ops/npt_nph.py]: Fixed in 65403cdfa1a810e645e9360ace3a7fc2e2332c77.

File: nvalchemi/dynamics/integrators/npt.py [hunk 2]
```diff
@@ -464,13 +447,14 @@ def post_update(self, batch: Batch) -> None:
         cells_inv = torch.linalg.inv_ex(batch.cell)[0].contiguous()
         KE = self._compute_ke(batch)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemi/dynamics/integrators/npt.py]: Fixed in 79300e89d619803843a69464fee77feba62ee905.

File: nvalchemi/dynamics/_ops/npt_nph.py [hunk 3]
```diff
@@ -91,6 +94,34 @@
...
+# ``__version__ == "0.3.1"`` until the next release bumps it, so
+# ``importlib.metadata.version`` cannot distinguish the two kernels.
+#
+# TODO: drop this flag, the bounded zero-buffer cache, and the call-site ``if``
+# in ``npt_barostat_half_step`` when the toolkit's minimum ops pin advances
+# above the legacy kernel.
+_NPT_BAROSTAT_LEGACY_API = "eta_dots" in inspect.signature(_npt_baro_half).parameters
+
+
+@lru_cache(maxsize=1)
```
Comment (laserkelvin) [inline nvalchemi/dynamics/_ops/npt_nph.py]: I don't really understand the impact of lru_cache on a single torch.tensor. Why is it done this way?

File: nvalchemi/dynamics/_ops/npt_nph.py [hunk 4]
```diff
@@ -494,27 +526,33 @@ def npt_barostat_half_step(
...
         _target_pressure_wp_array(target_pressure),
         wp.from_torch(volumes, dtype=scl_t),
         wp.from_torch(W, dtype=scl_t),
         wp.from_torch(kinetic_energy, dtype=scl_t),
         wp.from_torch(num_atoms_per_system, dtype=wp.int32),
-        wp.from_torch(eta_dots, dtype=scl_t),
-        wp.from_torch(dt, dtype=scl_t),
-    )
+    ]
+    if _NPT_BAROSTAT_LEGACY_API:
```
Comment (laserkelvin) [inline nvalchemi/dynamics/_ops/npt_nph.py]: If this is the only place that uses the flag, just add the condition here instead of defining a private variable in the module

File: test/dynamics/test_npt_nph.py
Comment (laserkelvin) [inline test/dynamics/test_npt_nph.py:1]: Do you mind adding a hard coded numerical regression test - the new test is just testing the API but not necessarily for correctness

**Review by laserkelvin — CHANGES_REQUESTED** [review]

File: test/dynamics/test_npt_nph.py
Comment (zubatyuk) [reply test/dynamics/test_npt_nph.py:1]: Added in 2d81d3c9f12072abc7357dba268c48e66a0e093d.

File: nvalchemi/dynamics/_ops/npt_nph.py (hunk 4 above)
Comment (zubatyuk) [reply nvalchemi/dynamics/_ops/npt_nph.py]: Fixed in 0bf55fd663fb3643402d6fc9c1f0dcb7ac4ea43f.

File: nvalchemi/dynamics/_ops/npt_nph.py (hunk 3 above)
Comment (zubatyuk) [reply nvalchemi/dynamics/_ops/npt_nph.py]: Switched to re-allocation on legacy path in 0bf55fd663fb3643402d6fc9c1f0dcb7ac4ea43f.

**Review by laserkelvin — APPROVED** [review]
